server/src/main/java/com/linecorp/centraldogma/server/ZooKeeperReplicationConfig.java

```diff
@@ -45,49 +63,56 @@
-     */
-    public ZooKeeperReplicationConfig(String connectionString, String pathPrefix,
+    @VisibleForTesting
+    ZooKeeperReplicationConfig(
+            int serverId, Map<Integer, ZooKeeperAddress> servers, String secret,
+            Map<String, String> additionalProperties,
             int timeoutMillis, int numWorkers, int maxLogCount, long minLogAgeMillis) {
-        this(connectionString, pathPrefix, Integer.valueOf(timeoutMillis),
+        this(Integer.valueOf(serverId), servers, secret, additionalProperties, Integer.valueOf(timeoutMillis),
              Integer.valueOf(numWorkers), Integer.valueOf(maxLogCount), Long.valueOf(minLogAgeMillis));
     }
 
     @JsonCreator
-    ZooKeeperReplicationConfig(@JsonProperty("connectionString") String connectionString,
-                               @JsonProperty("pathPrefix") String pathPrefix,
+    ZooKeeperReplicationConfig(@JsonProperty("serverId") @Nullable Integer serverId,
+                               @JsonProperty("servers")
+                               @JsonDeserialize(keyAs = Integer.class, contentAs = ZooKeeperAddress.class)
+                               @Nullable Map<Integer, ZooKeeperAddress> servers,
+                               @JsonProperty("secret") @Nullable String secret,
+                               @JsonProperty("additionalProperties")
+                               @JsonDeserialize(keyAs = String.class, contentAs = String.class)
+                               @Nullable Map<String, String> additionalProperties,
                                @JsonProperty("timeoutMillis") @Nullable Integer timeoutMillis,
                                @JsonProperty("numWorkers") @Nullable Integer numWorkers,
                                @JsonProperty("maxLogCount") @Nullable Integer maxLogCount,
                                @JsonProperty("minLogAgeMillis") @Nullable Long minLogAgeMillis) {
 
-        this.connectionString = requireNonNull(connectionString, "connectionString");
-        this.pathPrefix = requireNonNull(pathPrefix, "pathPrefix");
+        requireNonNull(servers, "servers");
+        this.serverId = serverId != null ? serverId : findServerId(servers);
+        checkArgument(this.serverId > 0, "serverId: %s (expected: > 0)", serverId);
 
-        int timeoutMillisValue = timeoutMillis != null ? timeoutMillis : DEFAULT_TIMEOUT_MILLIS;
-        if (timeoutMillisValue <= 0) {
-            timeoutMillisValue = DEFAULT_TIMEOUT_MILLIS;
-        }
+        this.secret = firstNonNull(secret, DEFAULT_SECRET);
+        checkArgument(!this.secret.isEmpty(), "secret is empty.");
+
+        servers.forEach((id, server) -> {
+            checkArgument(id > 0, "servers contains non-positive server ID: %s (expected: > 0)", id);
+        });
+        this.servers = ImmutableMap.copyOf(servers);
+
+        checkArgument(!this.servers.isEmpty(), "servers is empty.");
```

Comment:
`this.servers` is different from `servers`. I chose to check the emptiness of `this.servers` because `servers` may be mutable and modified concurrently from other thread.